src/UglyToad.PdfPig/DocumentLayoutAnalysis/TextEdgesExtractor.cs

```diff
@@ -0,0 +1,83 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using UglyToad.PdfPig.Content;
+using UglyToad.PdfPig.Geometry;
+
+namespace UglyToad.PdfPig.DocumentLayoutAnalysis
+{
+    /// <summary>
+    /// Text edges extractor. Text edges are where words have either there BoundingBox's left, right or mid coordinates aligned on the same vertical line.
+    /// <para>Useful to detect text columns, tables, justified text, lists, etc.</para>
+    /// </summary>
+    public class TextEdgesExtractor
+    {
+        /// <summary>
+        /// Functions used to define left, middle and right edges.
+        /// </summary>
+        private static readonly Tuple<string, Func<PdfRectangle, decimal>>[] edgesFuncs = new Tuple<string, Func<PdfRectangle, decimal>>[]
+        {
+            Tuple.Create<string, Func<PdfRectangle, decimal>>("left",   x => Math.Round(x.Left, 0)),                // use BoundingBox's left coordinate
+            Tuple.Create<string, Func<PdfRectangle, decimal>>("mid", x => Math.Round(x.Left + x.Width / 2, 0)),     // use BoundingBox's mid coordinate
+            Tuple.Create<string, Func<PdfRectangle, decimal>>("right",  x => Math.Round(x.Right, 0))                // use BoundingBox's right coordinate
+        };
+
+        /// <summary>
+        /// Get the text edges.
+        /// </summary>
+        /// <param name="pageWords">The words in the page.</param>
+        /// <param name="minimumElements">The minimum number of elements to define a text edge.</param>
+        public static Dictionary<string, List<PdfLine>> GetEdges(IEnumerable<Word> pageWords, int minimumElements = 4)
+        {
+            var cleanWords = pageWords.Where(x => !string.IsNullOrWhiteSpace(x.Text.Trim()));
+
+            ConcurrentDictionary<string, List<PdfLine>> dictionary = new ConcurrentDictionary<string, List<PdfLine>>();
+
+            Parallel.ForEach(edgesFuncs, f =>
+            {
+                dictionary.TryAdd(f.Item1, GetVerticalEdges(cleanWords, f.Item2, minimumElements));
+            });
+            return dictionary.ToDictionary(x => x.Key, x => x.Value);
```

Comment:
One thing it is perhaps worth doing here is using enum instead of string values for the key in the returned and internal dictionary, so an enum inside this class, `TextEdgesExtractor.EdgeType` like:

```
public enum EdgeType
{
   Left = 0,
   Mid = 1,
   Right = 2
}
```

This avoids people having to know up-front the correct case and terminology for the keys in this dictionary (e.g. is it `Mid`, `mid`, `MID`, `Middle`, ...).

Reply:
yes much better